채팅 기능 관련 Actor 구현

apptie opened the pull request:
## 📎 관련 Issue 번호
<!-- (필수) 해당 PR과 관련 있는 issue 번호를 입력해주세요. -->
- closed #3 
## 📄 작업 내용 요약
<!-- (필수) 작업한 내용을 간단히 요약해주세요. -->

### Actor 역할 판단 및 근거

- ChatChannelEntity : 도메인 모델
  - 채팅 채널의 핵심 비즈니스 개념 표현
  - Primary-Secondary에서 Primary 도메인 객체로 메시지 순서 보장, Secondary 관리 등 비즈니스 규칙을 캡슐화
  - WebSocket, DB 등 기술적 세부사항과 독립적
- ChatChannelReaderActor : 도메인 모델
  - Primary-Secondary에서 읽기 모델을 구현하는 Secondary 도메인 객체
  - Secondary로 읽기 전용 역할이라는 비즈니스 책임을 명확히 가짐
  - 특정 사용자의 채널 참여 상태를 나타내는 도메인 개념 (1:1로 매핑)
- ClientSessionActor(이전 ClientActor) : Inbound Adapter
  - ClientMessageSender를 통해 클라이언트에게 (웹소켓으로) 채팅 메시지를 전달하는 역할
  - WebSocket이라는 구체적인 기술과 도메인을 연결하는 일종의 브릿지
- NodeManagerActor(이전 SpawnClientActor) : Infrastructure Adapter
  - 물리적인 JVM에 단일로 존재해야 하는 종속적인 Actor를 생성하는 역할
  - 비즈니스 로직이 아닌 Actor 시스템 운영을 위해 존재
- MessageStorageActor(이전 DbActor) : Outbound Adapter
  - 영속화 로직(MessageRepository)을 호출하는 역할 
  - 별도의 DB Dispatcher를 사용하여 인프라 계층 분리가 명확
  - 영속화라는 기술적 관심사만을 담당

### 서로 다른 영역 간 Actor 메시지 관리 
- Domain 영역
  - Actor 간 상호작용을 위한 메시지를 일종의 포트로 취급 
    - Domain Actor에서 메시지를 주고받기 위해 필요한 인터페이스는 모두 port로 위치시킴 
    - 구분을 위해 Protocol 접미사 사용
- Domain 영역 외
  - Domain 영역이 아니라면 기존 방식대로 Actor 내 메시지 정의 
  - DB와 같이 Dispatcher의 성격이 다른 경우 외부 Port 구현체를 통해 메시지를 간접적으로 전달

### 기타
- BlockHound가 Spring WebFlux 관련 로직에서만 블로킹 로직을 감지하도록 설정 
- Actor 구조 구현에만 집중할 수 있도록 Repository 구현체는 Skeleton 구현만 추가

apptie committed:
docs: README.md BlockHound 설정 반영

apptie committed:
config: java.io.serialization 비활성화

apptie committed:
config: java.io.serialization 비활성화

apptie committed:
refactor: 패키지 구조 변경

apptie committed:
chore: .gitignore out 패턴 변경

apptie committed:
chore: Apache Pekko 테스트 의존성 추가

apptie committed:
fix: CborSerializable 마커 인터페이스가 도메인 인터페이스를 확장하도록 변경

apptie committed:
test: 스프링 부트 기본 테스트 삭제

apptie committed:
feat: 메시지 시퀀스 번호 생성 기능 추가

apptie committed:
feat: 채팅 메시지 도메인 모델 추가

- ChatMessage: 개별 메시지 Value Object
- ChatMessages: 메시지 컬렉션 관리

apptie committed:
feat: 채팅 메시지 Reader Actor 추가

apptie committed:
refactor: 패키지 구조 개선

- ActorCommand는 공통적으로 사용하는 마커 인터페이스이므로 domain/chat/common으로 이동
- 포트 역할을 수행하는 Protocol의 위치를 domain/chat/port/out으로 이동

apptie committed:
refactor: ChatChannelReaderActor 메시지 분리

- 포트 역할을 수행하는 메시지를 domain/chat/port/out/ChatChannelReaderProtocol로 분리

apptie committed:
feat: 채팅 메시지 Primary Entity 추가

- Cluster Sharding 기반 ChatChannelEntity 추가
- 채팅 메시지 순서 보장
- Secondary Reader 관리

apptie committed:
test: MockitoExtension 제거

apptie committed:
feat: 클라이언트 세션 관리 Actor 추가

- ClientMessageSender로 채팅 메시지 전달
- 세션 종료 처리

apptie committed:
feat: 물리적인 JVM의 Actor를 관리하는 Manager Actor 추가

- GuardianActor에서 NodeManagerActor spawn
- GuardianActor 생성 시 MessageStoragePort 주입
- Cluster가 필요한 테스트와 필요하지 않은 테스트 분리

apptie committed:
feat: 메시지 영속성을 담당하는 Actor 추가

- GuardianActor에서 MessageStorageActor spawn
- GuardianActor 생성 시 MessageRepository 주입
- EventStream 구독으로 저장/조회 이벤트 처리
- Actor Main Dispatcher와 분리된 persistence dispatcher 설정 추가
- BlockHound persistence-dispatcher 스레드를 블로킹 허용하도록 설정
- BlockHound 테스트 코드 추가

apptie committed:
feat: EventStream 기반 MessageStoragePort Adapter 추가

- ActorSystem EventStream으로 메시지 저장/조회 이벤트 발행
- ObjectProvider로 ActorSystem 순환참조 처리
- StoredEvent, LoadedHistoryEvent를 EventStream에 publish

apptie committed:
chore: 영속성 및 WebSocket 어댑터 skeleton 구현 추가

- JdbcMessageRepository stub
- WebSocketClientMessageSender stub
- Actor 시스템 의존성 해결을 위한 임시 구현

apptie committed:
refactor: NodeManagerActor 메서드 분리

apptie committed:
refactor: ChatChannelEntity 메서드 분리 및 메시지 네이밍 변경

apptie committed:
feat: 최근 메시지 동기화 기능 추가

- MessageRepository 메서드 추가
- MessageStoragePort 메서드 추가
- EventStream 이벤트 추가
- ChatChannelEntity Command 추가

apptie committed:
fix: ChatMessages 채팅 히스토리 동기화 방식 변경

- 생성자에서 동기화하지 않고 별도 메서드를 통해 동기화

apptie committed:
fix: ChatChannelEntity 생성 시 Port로부터 채팅 히스토리를 동기화하도록 변경

apptie committed:
remove: 불필요한 클래스 삭제

apptie committed:
feat: NodeManagerActor 등록 이벤트에 대한 EventStream 추가

seonWKim commented on pekko/src/main/java/com/tok/pekko/domain/chat/model/ChatMessages.java:
```diff
@@ -0,0 +1,90 @@
+        return copy;
+    }
+
+    public void add(ChatMessage message) {
```
유일한 entity에서 add를 호출하니 동기화는 필요가 없는걸까요??

apptie replied:
Actor 특성상 메시지를 순차적으로 하나씩 처리하기 때문에 락과 같은 동시성 기법을 적용하지 않고도 스레드 안정성이 보장되는 것으로 알고 있습니다 
ChatMessages를 관리하는 대상이 ChatChannelEntity와 ChatChannelReaderActor로 모두 Actor이기 때문에 동기화가 필요 없을 것이라고 판단했었습니다

seonWKim commented on pekko/src/main/java/com/tok/pekko/domain/chat/model/ChatChannelReaderActor.java:
```diff
@@ -0,0 +1,100 @@
+    private Behavior<ChatChannelReaderCommand> onRequestHistory(RequestHistory command) {
+        List<ChatMessage> history = this.messages.getHistory(command.messageSequence(), command.size());
+
+        if (history.isEmpty()) {
```
`ChatChannelReaderActor`에서  `messageStoragePort`를 직접 조회하는 이유가 궁금해요. `ChatChannelReaderActor`는 secondary라고 설명해주셨는데 primary를 건너뛰고 데이터를 읽는 로직처럼 보여서요.

apptie replied:
Primary-Secondary랑 맞지 않는 구조라고 생각은 했었습니다 

그런데 Primary가 자체적으로 ChatMessages 필드로 관리하고 메시지를 통해 Secondary에게 동기화를 수행하는데
그 과정이 Secondary가 요청하는 것이 아닌 Primary가 전파하는 것이다 보니
Secondary가 관리하고 있는 ChatMessages에 채팅 히스토리가 없다는 것은 Primary에게서도 없다는 걸 Secondary가 알고 있다고 생각했었습니다

그래서 Secondary가 Primary로 요청하는게 좀 어색하다고 느껴져서 이렇게 MessageStoragePort에게 직접 요청하도록 처리했습니다
확신은 없어서...이 부분에 대해 조언을 부탁드리고 싶습니다 

1. Primary가 전파를 하니 Secondary에게 동기화된 메시지를 토대로 Secondary가 Primary의 상태까지 추측할 수 있다는 제 생각이 자연스러운지 궁금합니다 
2. Secondary에서 데이터가 Primary에서까지 없는 걸 알면서도 일정한 패턴을 유지하기 위해 Primary에게 계속 요청을 하는게 좋을지 아니면 지금처럼 처리하는게 자연스러울지 궁금합니다

seonWKim replied:
>1. Primary가 전파를 하니 Secondary에게 동기화된 메시지를 토대로 Secondary가 Primary의 상태까지 추측할 수 있다는 제 생각이 자연스러운지 궁금합니다

자연스럽다고 생각합니다. 물론 동기화 로직이 잘 작성됐다는 전제하에요 

>2. Secondary에서 데이터가 Primary에서까지 없는 걸 알면서도 일정한 패턴을 유지하기 위해 Primary에게 계속 요청을 하는게 좋을지 아니면 지금처럼 처리하는게 자연스러울지 궁금합니다

Secondary의 컨셉을 어떻게 가져가시나요? 제가 생각했던 건 Primary의 read 부하를 덜기 위한 목적으로 생각했어요(어느 정도 복제 지연은 감수하구요). Secondary는 오로지 primary에서 싱크해주는 메시지를 전달받고, 전달받은 메시지를 토대로 클라이언트에게 응답하면 될거라 생각했습니다.

apptie replied:
저는 Secondary의 역할을 Primary에서 최신 메시지를 전달받아 메시지 순서를 유지하는 것까지만 고려하고 나머지 부분은 좀 더 자유롭게 가져가는걸 생각했습니다

그래서 Primary가 관리 하지 않는, DB에 저장되있는 데이터를 읽기위해서 Secondary가 Primary에 요청하거나 전파받는게 어색하게 느껴졌습니다

apptie replied:
그런데 첫 번째 질문에 해주신 답변 중 동기화 로직이 잘 되어 있다고 해주신 부분에서는 확신이 없습니다

지금 단순히 Actor 기반이니 막연하게 동기화를 잘 해주겠거니 하는 상황입니다
그렇다면 멘토님께서 말씀해주신대로 Secondary를 완전히 수동적으로 한다면 동기화 문제를 좀 배제할 수 있지 않을까하는 생각도 들었습니다

이 부분도 좀 더 깊게 생각해보도록 하겠습니다

seonWKim commented on pekko/src/main/java/com/tok/pekko/domain/chat/model/ChatMessages.java:
```diff
@@ -0,0 +1,90 @@
+        return copy;
+    }
+
+    public void add(ChatMessage message) {
```
원래는 동시성 이슈가 발생할 수 있으나 Actor 특성상 동시성 이슈가 발생하지 않음을 나타내는 marker annotation을 추가해서 이와 비슷한 메서드들에 어노테이션을 추가하면 어떨까요?

apptie replied:
이 생각은 전혀 하지 못했었는데 추가하도록 하겠습니다!

seonWKim commented on pekko/src/main/java/com/tok/pekko/domain/chat/port/in/ChatChannelProtocol.java:
```diff
@@ -0,0 +1,21 @@
+
+public interface ChatChannelProtocol {
+
+    interface ChatChannelEntityCommand extends ActorCommand { }
```
`ActorCommand`가 `CborSerializable`이 아닌 것 같은데 다른 노드로 메시지 송신이 가능한가요?

apptie replied:
찾아보니까 ActorCommand를 두고 CborSerializable가 이걸 구현하도록 한 뒤에 application.conf에서 명시하면 된다고는 해서 공통적으로 사용할 CborSerializable과 도메인 영역에서 사용할 ActorCommand로 분리할 수 있겠다 싶어서 시도했었습니다
 
TestKit으로 테스트 코드를 작성했을 때는 정상적으로 되기는 했었는데 
해당 코멘트 보고 조금 더 찾아보니 TestKit 단일 JVM에서만 동작하도록 하고 분산 환경에서의 직렬화는 보장을 안해준다는걸 알아버렸습니다...

SerializationTestKit이랑 직접 로컬에서 테스트해본 뒤 안되면 수정하도록 하겠습니다

apptie replied:
테스트해보니 정상적으로 직렬화가 안되서 메시지 송신이 불가능했습니다...
수정하도록 하겠습니다

seonWKim commented on pekko/src/main/java/com/tok/pekko/infrastructure/persistence/MessageStorageAdapter.java:
```diff
@@ -0,0 +1,56 @@
+@RequiredArgsConstructor
+public class MessageStorageAdapter implements MessageStoragePort {
+
+    private final ObjectProvider<ActorSystem<?>> actorSystemProvider;
```
ObjectProvider를 사용한 이유가 궁금해요

apptie replied:
ClusterConfig에서 ActorSystem을 빈으로 등록하는 과정에서 GuardianActor spawn 내부 과정에서 MessageStorageActor를 같이 spawn하는데 
MessageStorageActor가 내부적으로 EventStream을 사용해 ActorSystem을 빈으로 주입받아야 해서 순환 참조가 발생했습니다 

ActorSystem은 서버 초기화 과정에서 즉시 수행되고, 이후 MessageStorageAdapter는 채팅 메시지 관련해 외부와 소통해야 할 경우에만 동작하므로 이 시점에는 반드시 ActorSystem이 빈으로 초기화되었음을 보장한다고 생각해 ObjectProvider를 사용했습니다

seonWKim replied:
`ActorSystem` -> `GuardianActor` -> `MessageStorageActor` -> `ActorSystem` 이군요. 

`GuardianActor`에서 바로 `MessageStorageActor`를 생성해야하는 이유가 궁금해요.

apptie replied:
JVM 내에 단일로 존재해야하고 
ChatChannelEntity와 같은 domain actor가 준비되기전 설정이 모두 완료되어야하므로 GuardianActor에서 같이 ActorSystem 설정할 때 세팅되는것이 자연스러울 것이라고 판단했습니다

seonWKim replied:
> ChatChannelEntity와 같은 domain actor가 준비되기전 설정이 모두 완료

API 요청을 받기 전에만 설정이 완료되면 되는걸까요? 그럼 @PostConstruct, @DependsOn 등의 스프링 기능을 활용해서 bean 초기화 순서를 조율해보면 어떨까요? 그럼 위와 같은 circular dependency도 제거할 수 있을 것 같아요

apptie replied:
이해했습니다 
말씀해주신 방향대로 수정하도록 하겠습니다

apptie replied:
그런데 MesageStoreAdapter가 바로 MessageRepository를 호출하는 방향으로 간다면 해당 부분은 고려할 필요가 없어질 것 같습니다

나중에 이러한 순환참조가 생기는 경우 말씀하신 방법으로 처리하도록 하겠습니다

seonWKim commented on pekko/src/main/java/com/tok/pekko/domain/chat/model/ChatChannelEntity.java:
```diff
@@ -0,0 +1,147 @@
+        ChatMessage message = createChatMessage(command);
+
+        messages.add(message);
+        messageStoragePort.store(message);
```
store에 실패하면 어떻게 되나요?

apptie replied:
실패하는 경우 해당 채팅 메시지가 그대로 누락될 것 같습니다 
지금 구조상 메시지 영속화랑 세션으로 채팅 메시지 전파랑 별도로 동작하는 상황이므로 
누락된 채팅을 올린 클라이언트에서만 세션을 끊을 때 까지만 해당 메시지를 볼 수 있고 이후에는 완전히 사라지게 될 것 같습니다 

이에 대응하기 위해 단순하게 retry 로직을 추가하거나 트랜잭션 아웃박스 패턴 같은걸 고려할 생각입니다 
Pekko Projection이랑 엮어서 한 번 시도해볼 것 같습니다

apptie replied:
그런데 이와 별개로 Actor 간 메시지 손실에 대한 내용을 어디까지 고려해야 되는지도 고민이 되었습니다
채팅 서비스 특성상 채팅 메시지가 누락되면 사용자 경험에 굉장히 좋지 않을 것이라고 생각하고 있는데 
Actor가 분산 환경을 전제로 최대 한 번만 메시지를 전달한다는 특성을 가지고 있다보니 이에 대해서 어느 정도로 대응해야 할지가 좀 어려웠습니다 

Actor간 모든 메시지를 Reliable Delivery 같은걸 써야 하나 싶기도 하고 메시지 손실 방지를 위해 메일박스를 처리해야 할지 고민을 했었습니다
근데 너무 과한 것 같기도 하고 메일박스는 시도할 엄두조차 나지 않더라고요...
그래서 이와 관련해서 조언을 부탁드리고 싶습니다

seonWKim replied:
1. `messageStoragePort.store(message)` 
2. `messageStoragePort` 저장 성공 -> `ChatChannelEntity`로 메시지 전송 
3. `messages.add(message)`

요런 플로우는 어떨까요

apptie replied:
아예 영속화 이후 메시지를 동기화 시킨다면 누락될 가능성도 적어질것이고
만약 사용자가 메시지를 보냈을 때 영속화에 실패했다면 바로 실패했다는 결과를 얻을 수 있을테니 이전보다 핸들링이 편할 것 같습니다

이 방향으로 수정해보겠습니다! 감사합니다!

seonWKim commented on pekko/src/main/java/com/tok/pekko/domain/chat/model/ChatMessages.java:
```diff
@@ -0,0 +1,90 @@
+            throw new IllegalArgumentException("메시지는 null 일 수 없습니다.");
+        }
+
+        messages.addFirst(message);
```
`message`가 `messages`에 저장된 다른 `message`들보다 항상 앞선 sequence number을 가졌음을 보장할 수 있나요?

apptie replied:
sequence number는 ChatChannelEntity가 가지고 있는 MessageSequenceGenerator가 생성하도록 해서
특정 채널에 대한 ChatChannelEntity 생성 시 최초 생성이라면 1부터 시작하도록 하고 최초 생성이 아니라면 DB에서부터 가장 최신 sequence number를 가져올 것을 가정하고 작성했습니다

그래서 addFirst()로 앞에 넣기는 하지만 sequence number는 항상 더 큰 값을 가지도록 보장할 수 있다고 가정했습니다

apptie replied:
이 부분도 좀 고민이 많았는데 
그냥 DB에서 auto increment를 활용할까 싶었지만 기본적으로 비동기로 동작하는 Actor에서 DB insert 과정을 기다리는 블로킹이 추가되어도 괜찮을지가 고민되었습니다

임시방편으로 MessageSequenceGenerator으로 sequence number를 맞추도록 하기는 했지만 Passivation 같은 Actor 재시작을 전혀 고려하지 않기도 했고 
Split Brain 같은 분산 환경 또한 전혀 고려하지 않았습니다 
이런걸 생각하면 그냥 DB한테 id 만들라고 시키는게 나아보이긴 하는데 이 부분에 대해서도 조언을 부탁드리고 싶습니다

seonWKim replied:
요건 내일 조금 더 생각해보시죠. uuidv7도 고려해보세요

apptie replied:
uuid v7이나 snowflake도 생각해보기는 했었는데 과한게 아닌가 싶기도 하고 한번도 해보질 않아서 제외했었습니다

의존성 가져와서 사용하면 되기도 하고 각 채널마다 중복되도 되니 snowflake도 괜찮지 않나 싶기도 하고..말씀해주신대로 좀 더 고민해보도록 하겠습니다

seonWKim commented on pekko/src/main/java/com/tok/pekko/infrastructure/persistence/MessageStorageAdapter.java:
```diff
@@ -0,0 +1,56 @@
+
+@Component
+@RequiredArgsConstructor
+public class MessageStorageAdapter implements MessageStoragePort {
```
`MessageStorateAdapter`에서 `MessageRepository` 메서드를 호출하는게 아니라 별도 Actor를 호출하는 장점이 궁금해요.

apptie replied:
Actor에서 MessageStoragePort를 통해 MessageStorageAdapter를 호출하는데 
만약 MessageStorageAdapter에서 직접 MessageRepository를 호출하게 되면 블로킹 작업이 호출되어 비동기로 동작하는 Actor와 좀 어울리지 않는다고 생각했습니다 
그래서 별도 블로킹 처리용 스레드 풀에서 동작시키게 하고 싶었는데 직접 제가 관리하는 것보다는 Apache Pekko 설정으로 별도 Dispatcher를 생성해서 처리하는 것이 더 낫겠다고 판단해서 이렇게 구현했습니다

seonWKim replied:
>만약 MessageStorageAdapter에서 직접 MessageRepository를 호출하게 되면 블로킹 작업이 호출되어 비동기로 동작하는 Actor와 좀 어울리지 않는다고 생각했습니다

`MessageStorageActor`에서 `MessageRepository`를 호출하는데 요긴 괜찮다고 판단하신걸까요? 어디선가는 결국 동기 로직을 수행해야하는데, 비동기와 동기를 자연스럽게 연결하는 부분이 `MessageStorageAdapter`이지 아닐까 생각했습니다.

apptie replied:
MessageStorageActor는 dispatcher를 별도로 설정했으므로 괜찮다고 생각했었습니다

지금 프로젝트가 Actor 기반이니 MessageStorageAdapter보다는 단순 연결만 해주고 실제 동기 비동기 동작을 선택하는 것은 Actor쪽에서 하는게 더 자연스럽다고 생각한것도 있습니다

seonWKim replied:
>지금 프로젝트가 Actor 기반이니 MessageStorageAdapter보다는 단순 연결만 해주고 실제 동기 비동기 동작을 선택하는 것은 Actor쪽에서 하는게 더 자연스럽다고 생각한것도 있습니다 

뭐가 더 자연스러운지는 사람마다 의견이 나뉠 것 같네요 🤔 . 다만 여기서 actor를 사용하면 로직의 복잡도가 올라가고 성능적인 면(message hop 증가)에서도 불리한데 actor 사용이 적절할지는 의문이에요.

seonWKim replied:
>MessageStorageActor는 dispatcher를 별도로 설정했으므로 괜찮다고 생각했었습니다

지금 구조를 보니 actor는 1개, dispatcher는 5개의 스레드를 사용하는 것 같은데요, actor의 수가 한개라서 처리량은 증가하지 ㅇ않을 것 같습니다. 요건 혹시 의도하신걸까요??

apptie replied:
> 지금 구조를 보니 actor는 1개, dispatcher는 5개의 스레드를 사용하는 것 같은데요, actor의 수가 한개라서 처리량은 증가하지 않을 것 같습니다. 요건 혹시 의도하신걸까요??

처리량은 전혀 고려하지 않았습니다 
일차적으로 블로킹되는 로직을 호출하는 Actor의 Dispatcher를 분리하는 것을 목적으로 두었습니다

추가적으로 dispatcher가 5개의 스레드를 사용한다고 말씀해주셨는데 
application-dev.conf에서 persistence-dispatcher에 스레드를 10개로 고정했었고 
MessageStorageActor를 GuardianActor에 spawn할 때 persistence-dispacher를 지정해서 10개의 스레드를 사용하도록 의도했었는데 혹시 제가 설정을 잘못한 부분이 있었을까요?

seonWKim replied:
네넵, 10개로 설정되어있네요! 제가 잘못 본 것 같습니다.

apptie replied:
> 다만 여기서 actor를 사용하면 로직의 복잡도가 올라가고 성능적인 면(message hop 증가)에서도 불리한데 actor 사용이 적절할지는 의문이에요.

로직의 복잡도는 감수할만하고 생각했는데 메시지 홉 관련된 부분은 고려하지 못했습니다 
성능까지 악영향을 끼친다면 로직의 복잡도도 낮추기 위해 말씀해주신대로 Adapter에서 처리하는게 더 나을 거라는 생각이 듭니다 
이 부분도 수정하도록 하겠습니다

seonWKim commented on pekko/src/main/java/com/tok/pekko/infrastructure/persistence/actor/MessageStorageActor.java:
```diff
@@ -0,0 +1,127 @@
+                event -> new Store(event.message())
+        );
+
+        context.getSystem()
```
요 코드의 목적은 뭔가요??

apptie replied:
MessageStorageAdapter에서 MessageStorageActor로 메시지를 전달하기 위해서 EventStream을 구독하는 용도로 코드를 작성했습니다

seonWKim commented on pekko/src/main/java/com/tok/pekko/adapter/out/persistence/JdbcMessageRepository.java:
```diff
@@ -0,0 +1,27 @@
+import org.springframework.stereotype.Component;
+
+@Component
+public class JdbcMessageRepository implements MessageRepository {
```
infra와 adapter에 persistence 관련 로직을 분리해주신 이유가 궁금해요

apptie replied:
제 의도는
아직 JPA를 쓸지 JDBC를 쓸지 확실하게 정해지 못한 것 같아서 MessageRepository로 구현 기술을 분리할 수 있도록 했으며 
MessageStorageAdapter는 MessageStoragePort를 구현하면서 EventStream으로 다른 Actor에게 메시지를 전달하는 브릿지 역할만 수행하는, Apache Pekko 기술에 강하게 종속되어 있다고 판단해 infra 계층이라고 생각했었습니다  

그런데 JdbcMessageRepository는 지금보니 infra 구현체인데 adapter에 제가 잘못 배치한 것 같습니다 
인터페이스랑 구현체랑 같이 있는게 뭔가 어색해서 adapter로 옮겼는데 여러모로 찾아보니 아닌 것 같습니다
이 부분 수정하도록 하겠습니다

seonWKim commented on pekko/src/main/java/com/tok/pekko/application/actor/NodeManagerActor.java:
```diff
@@ -0,0 +1,176 @@
+import org.apache.pekko.cluster.sharding.typed.javadsl.ClusterSharding;
+import org.apache.pekko.cluster.sharding.typed.javadsl.EntityRef;
+
+public class NodeManagerActor extends AbstractBehavior<NodeManagerActorCommand> {
```
>NodeManagerActor(이전 SpawnClientActor) : Infrastructure Adapter
물리적인 JVM에 단일로 존재해야 하는 종속적인 Actor를 생성하는 역할
비즈니스 로직이 아닌 Actor 시스템 운영을 위해 존재

지민님이 생각하시는 비즈니스 로직의 범주가 어떻게 되나요?

apptie replied:
현재 구조에서는 상당히 좁은 범위라고 생각되기는 하지만 채팅 메시지를 관리하는 것이 비즈니스 로직의 범위라고 생각했습니다 

그래서 ChatChannelEntity와 ChatChannelReaderActor를 제외한 나머지 Actor는 전부 도메인 영역이 아닌 다른 영역에 배치시키는 식으로 진행했었습니다

apptie committed:
fix: ActorCommand 제거

- application.conf에 직렬화를 명시한 CborSerializable가 ActorCommand를 확장하더라도 정상적으로 직렬화되지 않음
- ActorCommand 삭제 후 CborSerializable를 모든 영역에서 공통적으로 사용해야 하므로 외부 common 패키지로 이동

apptie committed:
refactor: 패키지 구조 개선

- domain 영역의 port를 사용하거나 구현하는 경우 adapter 영역으로 설정
- adapter 영역에서만 사용하는 경우 infrastructure 영역으로 설정

apptie committed:
feat: Actor 기반 thread-safe한 메서드를 표현하는 marker annotation 추가

- Actor 모델이 thread-safety를 보장하는 메서드를 명시적으로 표현
- ChatMessages 클래스의 메서드들에 적용

apptie committed:
fix: 채팅 메시지 동기화 로직 개선

- 영속화에 성공한 메시지만을 동기화하도록 변경

apptie committed:
refactor: DB 로직 호출 과정 변경

- Actor 메시지 전달 방식을 제거하고 Adapter가 직접 Repository 호출
- Actor message hop 제거
- 로직 복잡도 감소

apptie committed:
refactor: 채팅 메시지 시퀀스 부여 로직 변경

- 기존 단순 시퀀스 값 증가 방식은 메시지 순서 보장을 확신할 수 없음
- Snowflake 알고리즘으로 타임스탬프 기반 순서 보장 가능한 시퀀스 생성

apptie committed:
refactor: 채팅 히스토리 조회 책임 이동

- ChatChannelReaderActor에서 ChatChannelEntity로 이동
- ChatChannelReaderActor의 MessageStoragePort 의존성 제거

apptie committed:
fix: Actor에서 채팅 히스토리 조회 책임 제거

- 채팅 히스토리는 Actor에서 별도로 관리하지 않음
- 채팅 히스토리 조회는 별도 컴포넌트로 구현 예정

apptie committed:
refactor: 패키지 구조 개선

- infrastructure 패키지 제거
- global 패키지 추가
- application 패키지 추가

apptie committed:
feat: ChatMessages 동기화 시 중복 메시지 제거 기능 추가

- messageId 기준 중복 제거 로직 구현
- 중복 시 새로운 메시지(Primary 데이터) 우선 선택

apptie committed:
feat: HeartBeat를 통한 메시지 동기화 기능 추가

- ChatChannelReaderActor에 30초 간격 HeartBeat 타이머 구현
- Entity와 Reader 간 RequestSyncMessages / DeliverSyncMessages 메시지 추가

apptie committed:
refactor: Actor 기반의 관리 방식을 Application Service 방식으로 변경

- Actor 내부 spawn에서 외부 Service 기반 관리로 변경
- ActorSystem에 SpawnProtocol 적용
- Ask 패턴을 통해 외부에서 Actor Spawn

seonWKim commented on pekko/src/main/java/com/tok/pekko/application/actor/SessionActorManagementService.java:
```diff
@@ -0,0 +1,114 @@
+                .thenCompose(clientSession -> spawnChatChannelReader(chatChannel, clientSession, userId, channelId))
+                .thenAccept(
+                        chatChannelReader -> {
+                            localChatChannelReaders.put(new NodeReaderKey(channelId, userId), chatChannelReader);
```
terminateSession()이 호출되지 않는 상황은 없을까요?

apptie replied:
WebSocketHandler에서 WebSocketSession을 핸들링할 때 Mono<Void>를 반환하면서 doFinally()를 활용해 terminateSession을 항상 호출할 수 있을 것이라 생각하고 있습니다

f-lab-devvy reviewed (approved):
리뷰 범위가 지나치게 넓어서 우선 요 PR은 머지하고 진행할게요. 이후 리뷰하면서 놓친 부분들 코멘트 남겨두겠습니다.

seonWKim commented on pekko/src/test/java/com/tok/pekko/adapter/out/persistence/MessageStorageAdapterTest.java:
```diff
@@ -0,0 +1,372 @@
+    }
+
+    @Test
+    void findRecentMessages_호출_시_호출자_스레드를_블로킹하지_않는다() throws InterruptedException {
```
이 테스트가 호출자 스레드를 블로킹하지 않음을 어떻게 확인하는건가요??

apptie replied:
해당 테스트에 @BeforeEach로 BlockHound.install()을 통해 Spring WebFlux 관련 스레드에 대해서는 블로킹을 감지하도록 했습니다 

블로킹 여부를 확인할 메서드를 Mono.fromRunnable()로 실행시키는데 이 경우 스레드 네이밍이 parallel로 시작하고 이를 BlockHoundTestInstallUtils.install에서 감시하도록 했기 때문에 블로킹 여부를 체크할 수 있었습니다
